src_feedback/elf.rs

```diff
@@ -0,0 +1,666 @@
+    st_info: u8,		/* Type and binding attributes */
+    st_other: u8,		/* No defined meaning, 0 */
+    st_shndx: Elf64_Half,	/* Associated section index */
+    st_value: Elf64_Addr,	/* Value of the symbol */
+    st_size: Elf64_Xword,	/* Associated symbol size */
+}
+
+#[allow(dead_code)]
+impl Elf64_Sym {
+    fn get_type(&self) -> u8 {
+	self.st_info & 0xf
+    }
+
+    fn is_undef(&self) -> bool {
+	self.st_shndx == SHN_UNDEF
+    }
+}
+
+#[allow(dead_code)]
+#[repr(C)]
+struct Elf64_Rel {
+    r_offset: Elf64_Addr,	/* Location at which to apply the action */
+    r_info: Elf64_Xword,	/* index and type of relocation */
+}
+
+#[allow(dead_code)]
+#[repr(C)]
+struct Elf64_Rela {
+    r_offset: Elf64_Addr,	/* Location at which to apply the action */
+    r_info: Elf64_Xword,	/* index and type of relocation */
+    r_addend: Elf64_Sxword,	/* Constant addend used to compute value */
+}
+
+#[allow(dead_code)]
+#[repr(C)]
+union Elf64_Dyn_un{
+    d_val: Elf64_Xword,
+    d_ptr: Elf64_Addr,
+}
+
+#[allow(dead_code)]
+#[repr(C)]
+struct Elf64_Dyn {
+    d_tag: Elf64_Sxword,	/* entry tag value */
+    d_un: Elf64_Dyn_un,
+}
+
+
+fn read_u8(file: &mut File, off: u64, size: usize) -> Result<Vec<u8>, Error> {
+    let mut buf = vec![0; size];
+
+    file.seek(SeekFrom::Start(off))?;
+    file.read_exact(buf.as_mut_slice())?;
+
+    Ok(buf)
+}
+
+fn read_elf_header(file: &mut File) -> Result<Elf64_Ehdr, Error> {
+    const DSZ: usize = mem::size_of::<Elf64_Ehdr>();
+    let mut buf = Box::new([0_u8; DSZ]);
+
+    let buf_m: &mut [u8; DSZ] = buf.borrow_mut();
+    file.read_exact(buf_m)?;
+
+    let ehdr: Box<Elf64_Ehdr> = unsafe {
+	// A complicated type casting hacking!!
+	let ehdr_raw_ptr = (Box::leak(buf) as *mut u8) as *mut Elf64_Ehdr;
+	Box::from_raw(ehdr_raw_ptr)
+    };
+
+    Ok(*ehdr)
+}
+
+fn read_elf_sections(file: &mut File, ehdr: &Elf64_Ehdr) -> Result<Vec<Elf64_Shdr>, Error> {
+    const HDRSIZE: usize = mem::size_of::<Elf64_Shdr>();
+    let off = ehdr.e_shoff as usize;
+    let num = ehdr.e_shnum as usize;
+
+    let mut buf = read_u8(file, off as u64, num * HDRSIZE)?;
+
+    let shdrs: Vec<Elf64_Shdr> = unsafe {
+	let shdrs_ptr = buf.as_mut_ptr() as *mut Elf64_Shdr;
+	buf.leak();
+	Vec::from_raw_parts(shdrs_ptr, num, num)
+    };
+    Ok(shdrs)
+}
+
+fn read_elf_program_headers(file: &mut File, ehdr: &Elf64_Ehdr) -> Result<Vec<Elf64_Phdr>, Error> {
+    const HDRSIZE: usize = mem::size_of::<Elf64_Phdr>();
+    let off = ehdr.e_phoff as usize;
+    let num = ehdr.e_phnum as usize;
+
+    let mut buf = read_u8(file, off as u64, num * HDRSIZE)?;
+
+    let phdrs: Vec<Elf64_Phdr> = unsafe {
+	let phdrs_ptr = buf.as_mut_ptr() as *mut Elf64_Phdr;
+	buf.leak();
+	Vec::from_raw_parts(phdrs_ptr, num, num)
+    };
+    Ok(phdrs)
+}
+
+fn read_elf_section_raw(file: &mut File, section: &Elf64_Shdr) -> Result<Vec<u8>, Error> {
+    read_u8(file, section.sh_offset as u64, section.sh_size as usize)
```

Comment:
sh_offset is a u64 so do we need this cast?

Reply:
No